outlines/text/generate/sequence.py

```diff
@@ -0,0 +1,259 @@
+from typing import List, Optional, Tuple, Union
+
+import numpy as np
+from numpy.random import Generator
+from numpy.typing import NDArray
+
+
+class Sequence:
+    """Represents a sequence generation method."""
+
+    def __init__(self, model, max_tokens: Optional[int] = None):
+        """Create a `Sequence` instance.
+
+        Parameters
+        ----------
+        model
+            The instance of the model used to generate next-token probabilities.
+        max_tokens
+            The maximum number of tokens that will be generated if no termination
+            condition is met.
+
+        """
+        self.model = model
+        self.max_tokens = max_tokens
+
+    def is_finished(self, token_ids: NDArray[np.int64]) -> NDArray[np.bool_]:
+        """Determine whether we should stop the generation."""
+        raise NotImplementedError(
+            "`Sequence.is_finished` must be implemented by subclasses."
+        )
+
+    def postprocess_completions(self, completions: List[str]) -> List[str]:
+        return completions
+
+    def step(
+        self,
+        rng: Generator,
+        token_ids: NDArray[np.int64],
+        attention_mask: NDArray[np.int64],
+        samples: int = 1,
+    ) -> Tuple[NDArray[np.int64], NDArray[float]]:
+        """Generate one or several tokens that complete the input sequence.
+
+        The sampling step consists in using a model to generate next token
+        probabilities and then sample one or several new tokens from a
+        categorical distribution parametrized by these probabilities.
+
+        Parameters
+        ----------
+        rng
+            NumPy random number Generator instance
+        token_ids
+            The token ids passed as an input to the model, of shape `batch_shape
+            + (num_tokens,)`.  The generated token is a random variable
+            parametrized by these token ids.
+        samples
+            The number of samples to take using the generated next token probabilities.
+
+        Returns
+        -------
+        A tuple with an array of shape `new_batch_shape + (num_tokens,)`that
+        contains the completion (input token ids and generated token ids) and an
+        array of shape `new_batch_shape + (vocab_size,)` that contains the next
+        token probabilities. `new_batch_shape` is computed by removing dimensions
+        of size one in `(samples,) + batch_shape`.
+
+        """
+        num_input_dims = token_ids.ndim
+        probs = self.model(token_ids, attention_mask)
+
+        # Sample `samples` new tokens
+        vocab_size = probs.shape[-1]
+        next_token_ids = vectorized_random_choice(
+            rng, np.arange(vocab_size), probs, samples
```

Comment:
We might want to consider how we can avoid creating an array like `np.arange(vocab_size)` if possible.  If `vectorized_random_choice` is only indexing that array, then we should be able to in at least this case, no?

Reply:
Good catch!